keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1347 @@
+      requirements:
+      - onExitCode:
+          notInRange:
+          - minInclusive: 40
+            maxInclusive: 50
+```
+
+Note that, when no rule specified in `backoffPolicy` matches the pod failure
+the default handling of pod failures applies - the counter of pod failures
+is incremented and checked against the `backoffLimit`
+(see: [JobSpec API](#jobspec-api)]).
+
+#### Story 2
+
+As a service provider that offers computational resources to researchers I would like to
+have a mechanism which terminates jobs for which pods are failing due to user errors,
+but allows infinite retries for infrastructure errors (pod terminations by kubelet,
+kube-scheduler, abnormal node termination, etc).
+I do not have knowledge or influence over the executable that researchers run,
+so I don't know beforehand which exit codes they might return.
+
+The following Job configuration could be a good starting point to satisfy
+my needs:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: main-job-container
+        image: job-image
+        command: ["./program"]
+      - name: monitoring-job-container
+        image: job-monitoring
+        command: ["./monitoring"]
+  backoffLimit: 3
+  backoffPolicy:
+    rules:
+    - action: Ignore
+      requirements:
+      - onPodCondition:
+        - key: Status
+          operator: In
+          values:
+          - True
+        - key: Type
+          operator: In
+          values:
+          - Preempted
+          - Evicted
+```
+
+Note that, in this case the user supplies a list of Pod condition type values
+which are interpreted as infrastructure failures. This approach is likely to
+require an iterative process to review and extend of the list.
+
+### Notes/Constraints/Caveats (Optional)
+
+Alternative versions of the API to define requirements on exit codes and on
+pod status have been proposed. We discuss them here based on the following
+example: let's assume we want to fail a job if the container exit code
+(for "main-container") is in range [1,10] or [100,127] . Also, we want
+to ignore the pod failure if the pod condition type is "Preempted" or "Evicted".
+
+In the current API we can formulate this policy as follows:
+
+```yaml
+ backoffPolicy:
+   rules:
+   - action: Terminate
+     requirements:
+     - onExitCodes:
+         containerName: main-container
+         inRange:
+         - minInclusive: 1
+           maxInclusive: 10
+         - minInclusive: 100
+           maxInclusive: 127
+   - action: Ignore
+     requirements:
+     - onPodCondition:
+       - key: Status
+         operator: In
+         values:
+         - True
+       - key: Type
+         operator: In
+         values:
+         - Preempted
+         - Evicted
+```
+
+Alternative: Source-based
+
+This approach unifies the API for exit codes and pod conditions using the
+concept of an operator:
+
+```yaml
+ backoffPolicy:
+   rules:
+   - action: Terminate
```

Comment:
are actions ordered?

Reply:
The rules (which are containers for a single action and a list of requirements) and ordered and processed from top to bottom.